### DISCUSSION TITLE
Somehow unsorted questions related to Zeptoforth

### DISCUSSION THREAD

**Discussion by mcctuxic** [tabemann/zeptoforth#113] [General]
Hi Travis,

some questions are in my head... :)

Is forth and especially zeptoforth homoiconic?

Since I am new to zeptoforth, I still have no clue, how to evaluate the amount of memory (all sorts) which
are available to a certain point of time. An example:
The word 'unused' gave my the following output on my Pi Pico W (no WiFi firmware currently):
```
unused 
flash dictionary free:     870656
main task dictionary free: 161228
main task stack free:      504
main task rstack free:     892 ok
```
504 bytes and 892 bytes free on both stacks.../sounds/ like I would soon run out of memory...for
a nerd, whose PC has 16GB of RAM...
How can I develope a better "feeling" for all related amounts of something, when programing zeptoforth
on a Pico W?

Did you ever "kill" a flash chip by writing too often to it? Are you still using your "my first Pi Pico"?  :)
(not counting software accidents, which errornously "hammer" on the flash a couple of thousand times per second.

Beside the RP2040 datasheet is there any good documentation about this MCU freely available?

Have a nice weekend! :)

Cheers!
Tuxic

**Comment by mcctuxic** [comment]
Sorry, forgot one question:

When done with writing to the SDcard, what is the zeptoforths equivalent of "umount" or "eject safely" ?

Cheers!
Tuxic

**Comment by tabemann** [comment]
On Thu, Jul 11, 2024 at 11:39 AM mcctuxic ***@***.***> wrote:

> Hi Travis,
>
> some questions are in my head... :)
>
> Is forth and especially zeptoforth homoiconic?
>

Yes, Forth is homoiconic, and zeptoforth is a Forth. :)


> Since I am new to zeptoforth, I still have no clue, how to evaluate the
> amount of memory (all sorts) which
>
> are available to a certain point of time. An example:
> The word 'unused' gave my the following output on my Pi Pico W (no WiFi
> firmware currently):
>
> unused
> flash dictionary free:     870656
> main task dictionary free: 161228
> main task stack free:      504
> main task rstack free:     892 ok
>
> 504 bytes and 892 bytes free on both stacks.../sounds/ like I would soon
> run out of memory...for
> a nerd, whose PC has 16GB of RAM...
>

Oh, that is plenty of space for most applications. Remember that the main
task's data stack starts out at 512 bytes and its return stack starts out
at 1024 bytes.


> How can I develope a better "feeling" for all related amounts of
> something, when programing zeptoforth
> on a Pico W?
>

A good sense is based on that the RP2040 comes with 264 KB of RAM,
excluding the XIP cache, which rp2040_big builds make 172 KB available to
the user (the rest is used for the kernel and the flash dictionary index
hash table), and that most boards come with 2 MB of flash, of which the
rp2040_big builds make 1472 KB available for the flash dictionary.

Did you ever "kill" a flash chip by writing too often to it? Are you still
> using your "my first Pi Pico"? :)
> (not counting software accidents, which errornously "hammer" on the flash
> a couple of thousand times per second.
>

I have had two RP2040 boards die on me, and none of them were, to my
knowledge, were due to wearing out the flash.


> Beside the RP2040 datasheet is there any good documentation about this MCU
> freely available?
>
> Have a nice weekend! :)
>
> Cheers!
> Tuxic
>

The main documentation I've seen is the RP2040 datasheet and the source
code for the RP2040's HAL.

Travis

**Comment by tabemann** [comment]
On Thu, Jul 11, 2024 at 11:53 AM mcctuxic ***@***.***> wrote:

> Sorry, forgot one question:
>
> When done with writing to the SDcard, what is the zeptoforths equivalent
> of "umount" or "efect safely" ?
>
> Cheers!
> Tuxic
>

All FAT32 filesystem objects have a FAT32::FLUSH method that ensures that
all updates are written out to media. When using the current FAT32
filesystem, this can be done with FAT32-TOOLS::CURRENT-FS@ FAT32::FLUSH .

Travis

**Comment by mcctuxic** [comment]
On 07/11 10:35, tabemann wrote:
> On Thu, Jul 11, 2024 at 11:39 AM mcctuxic ***@***.***> wrote:
> 
> > Hi Travis,
> >
> > some questions are in my head... :)
> >
> > Is forth and especially zeptoforth homoiconic?
> >
> 
> Yes, Forth is homoiconic, and zeptoforth is a Forth. :)

:):):):):):):):):):):):):):):):):):):):):):):):):):):):):):):):):):):)

OH YEAH! Zeptoforth is the most forthy programming language under all
programming languages. Praised be the Forth, praised be its inkarnation Zeptoforth!
And see - everything went good after that! 

:):):):):):):):):):):):):):):):):):):):):):):):):):):):):):):):):):):)

 
> 
> > Since I am new to zeptoforth, I still have no clue, how to evaluate the
> > amount of memory (all sorts) which
> >
> > are available to a certain point of time. An example:
> > The word 'unused' gave my the following output on my Pi Pico W (no WiFi
> > firmware currently):
> >
> > unused
> > flash dictionary free:     870656
> > main task dictionary free: 161228
> > main task stack free:      504
> > main task rstack free:     892 ok
> >
> > 504 bytes and 892 bytes free on both stacks.../sounds/ like I would soon
> > run out of memory...for
> > a nerd, whose PC has 16GB of RAM...
> >
> 
> Oh, that is plenty of space for most applications. Remember that the main
> task's data stack starts out at 512 bytes and its return stack starts out
> at 1024 bytes.

Ok...that looks more assuring. I need to keep in mind, that it was
possible to run a complete operating system and a BASIC interpreter
on 64k main memory in the golden era of homecomputers.
By the way: The Jupiter Ace was a kind of Sinclair ZX running Forth
instead of Basic. It was SO much ahead of time, that it was not
a market hit...unfortunately.
 
> > How can I develope a better "feeling" for all related amounts of
> > something, when programing zeptoforth
> > on a Pico W?
> >
> 
> A good sense is based on that the RP2040 comes with 264 KB of RAM,
> excluding the XIP cache, which rp2040_big builds make 172 KB available to
> the user (the rest is used for the kernel and the flash dictionary index
> hash table), and that most boards come with 2 MB of flash, of which the
> rp2040_big builds make 1472 KB available for the flash dictionary.

I simple need to get used to it and to get rid of this 16GB-feeling in
my head... :)

 
> Did you ever "kill" a flash chip by writing too often to it? Are you still
> > using your "my first Pi Pico"? :)
> > (not counting software accidents, which errornously "hammer" on the flash
> > a couple of thousand times per second.
> >
> 
> I have had two RP2040 boards die on me, and none of them were, to my
> knowledge, were due to wearing out the flash.

THAT sounds very assuring! 
Especially when I think how often you need to flash - you are the author
and developer of zeptoforth!

And I...I am still hasitating to write to flash...
;)

 
> > Beside the RP2040 datasheet is there any good documentation about this MCU
> > freely available?
> >
> > Have a nice weekend! :)
> >
> > Cheers!
> > Tuxic
> >
> 
> The main documentation I've seen is the RP2040 datasheet and the source
> code for the RP2040's HAL.

The HAL seems to very good source to me ! That is the right "food" for
the coming weekend!

> Travis

Thanks a lot for the valuable info, Travis!

Cheers!
Tuxic

**Comment by mcctuxic** [comment]
On 07/11 10:38, tabemann wrote:
> On Thu, Jul 11, 2024 at 11:53 AM mcctuxic ***@***.***> wrote:
> 
> > Sorry, forgot one question:
> >
> > When done with writing to the SDcard, what is the zeptoforths equivalent
> > of "umount" or "efect safely" ?
> >
> > Cheers!
> > Tuxic
> >
> 
> All FAT32 filesystem objects have a FAT32::FLUSH method that ensures that
> all updates are written out to media. When using the current FAT32
> filesystem, this can be done with FAT32-TOOLS::CURRENT-FS@ FAT32::FLUSH .
> 
> Travis

There is a "hook" (the "init" mechanism) one can utilize to a customized
boot sequence.
Is there something similiar for a "clean shutdown/reboot" (that is for 
everything reboot/shutdown-similiar, which is intentionally)?
In that case, it would be possible to hook a "clean eject" into it...

Cheers!
Tuxic

**Comment by tabemann** [comment]
On Thu, Jul 11, 2024 at 8:27 PM mcctuxic ***@***.***> wrote:

> On 07/11 10:35, tabemann wrote:
> > On Thu, Jul 11, 2024 at 11:39 AM mcctuxic ***@***.***> wrote:
> >
> > > Hi Travis,
> > >
> > > some questions are in my head... :)
> > >
> > > Is forth and especially zeptoforth homoiconic?
> > >
> >
> > Yes, Forth is homoiconic, and zeptoforth is a Forth. :)
>
> :):):):):):):):):):):):):):):):):):):):):):):):):):):):):):):):):):):)
>
> OH YEAH! Zeptoforth is the most forthy programming language under all
> programming languages. Praised be the Forth, praised be its inkarnation
> Zeptoforth!
> And see - everything went good after that!
>
> :):):):):):):):):):):):):):):):):):):):):):):):):):):):):):):):):):):)
>


LOL! To be completely honest, zeptoforth is not a very traditional Forth,
with things like liberal use of local variables and whatnot.



> >
> > > Since I am new to zeptoforth, I still have no clue, how to evaluate the
> > > amount of memory (all sorts) which
> > >
> > > are available to a certain point of time. An example:
> > > The word 'unused' gave my the following output on my Pi Pico W (no WiFi
> > > firmware currently):
> > >
> > > unused
> > > flash dictionary free: 870656
> > > main task dictionary free: 161228
> > > main task stack free: 504
> > > main task rstack free: 892 ok
> > >
> > > 504 bytes and 892 bytes free on both stacks.../sounds/ like I would
> soon
> > > run out of memory...for
> > > a nerd, whose PC has 16GB of RAM...
> > >
> >
> > Oh, that is plenty of space for most applications. Remember that the main
> > task's data stack starts out at 512 bytes and its return stack starts out
> > at 1024 bytes.
>
> Ok...that looks more assuring. I need to keep in mind, that it was
> possible to run a complete operating system and a BASIC interpreter
> on 64k main memory in the golden era of homecomputers.
> By the way: The Jupiter Ace was a kind of Sinclair ZX running Forth
> instead of Basic. It was SO much ahead of time, that it was not
> a market hit...unfortunately.
>


The Jupiter Ace also had one *kilobyte* of RAM, which was, well,
restrictive, even at the time.



> > > How can I develope a better "feeling" for all related amounts of
> > > something, when programing zeptoforth
> > > on a Pico W?
> > >
> >
> > A good sense is based on that the RP2040 comes with 264 KB of RAM,
> > excluding the XIP cache, which rp2040_big builds make 172 KB available to
> > the user (the rest is used for the kernel and the flash dictionary index
> > hash table), and that most boards come with 2 MB of flash, of which the
> > rp2040_big builds make 1472 KB available for the flash dictionary.
>
> I simple need to get used to it and to get rid of this 16GB-feeling in
> my head... :)
>


One key thing when doing embedded programming is you quickly realize just
how big a kilobyte really is.



> > Did you ever "kill" a flash chip by writing too often to it? Are you
> still
> > > using your "my first Pi Pico"? :)
> > > (not counting software accidents, which errornously "hammer" on the
> flash
> > > a couple of thousand times per second.
> > >
> >
> > I have had two RP2040 boards die on me, and none of them were, to my
> > knowledge, were due to wearing out the flash.
>
> THAT sounds very assuring!
> Especially when I think how often you need to flash - you are the author
> and developer of zeptoforth!
>
> And I...I am still hasitating to write to flash...
> ;)
>


Don't be afraid to write to flash, and frankly, if you're working with
Pico's they are practically expendable with how cheap they are.



> > > Beside the RP2040 datasheet is there any good documentation about this
> MCU
> > > freely available?
> > >
> > > Have a nice weekend! :)
> > >
> > > Cheers!
> > > Tuxic
> > >
> >
> > The main documentation I've seen is the RP2040 datasheet and the source
> > code for the RP2040's HAL.
>
> The HAL seems to very good source to me ! That is the right "food" for
> the coming weekend!
>
> > Travis
>
> Thanks a lot for the valuable info, Travis!
>
> Cheers!
> Tuxic


 The thing about the HAL, though, is that it assumes a non-multitasking
system, while zeptoforth by its very nature is multitasking.

Travis
